src/components/HardwareWalletModal.js

```diff
@@ -71,39 +60,48 @@ const onConfirm = async (maker, address, path, closeModal) => {
   closeModal();
 };
 
-function LedgerAddresses({ onClose, isLedgerLive }) {
-  const [addresses, setAddresses] = useState([]);
+const TREZOR_PATH = "44'/60'/0'/0/0";
+const LEDGER_LIVE_PATH = "44'/60'/0'";
+const LEDGER_LEGACY_PATH = "44'/60'/0'/0";
+const DEFAULT_ACCOUNTS_PER_PAGE = 5;
+
+function HardwareWalletModal({ onClose, type, isLedgerLive }) {
+  const [addressList, setAddressList] = useState([]);
   const [selectedAddress, setSelectedAddress] = useState('');
+  const [accountCb, setAccountCb] = useFn(() => null);
   const { maker } = useMaker();
-  const { showSimpleByType } = useModal();
-  const path = isLedgerLive ? LEDGER_LIVE_PATH : LEDGER_LEGACY_PATH;
+  const path =
+    type === AccountTypes.TREZOR
+      ? TREZOR_PATH
+      : isLedgerLive
+      ? LEDGER_LIVE_PATH
+      : LEDGER_LEGACY_PATH;
 
   const walletAddresses = useMakerState(maker =>
     maker.addAccount({
-      type: AccountTypes.LEDGER,
+      type,
       path,
       accountsOffset: 0,
       accountsLength: DEFAULT_ACCOUNTS_PER_PAGE,
-      choose: addresses => {
+      choose: async (addresses, cb) => {
         const addressBalancePromises = addresses.map(address =>
           addMkrAndEthBalance({
             address,
-            type: AccountTypes.LEDGER
+            type
           })
         );
-        return Promise.all(addressBalancePromises).then(addressBalances =>
-          setAddresses(addressBalances)
-        );
+        setAddressList(await Promise.all(addressBalancePromises));
+        setAccountCb(async address => await cb(null, address));
```

Comment:
if this change gives you grief, could you try calling `onConfirm` with a 0 ms timeout like 
```js
setTimeout(() => onConfirm(maker, selectedAddress, onClose, type), 0);
```

test a hunch of mine 🙂

Reply:
Wow, the timeout worked, a very good hunch! The `cb()` is created [here](https://github.com/makerdao/dai-plugin-trezor-web/blob/6363b6d31f7280e167319173a67490658b2a0d60/src/index.js#L30) in the trezor-plugin. Tbh, the async/await worked where not including it didn't so that's why I went with it.

Reply:
So `cb()` adds an account to the sdk when the `await` on the promise it resolves is settled and the code below it executes, but since that `await` is put on the task queue, it is only run once the call stack is empty. When we were calling `useAccountWithAddress` after `cb()` here, those instructions were put on the call stack, so they would always be executed before the account had been added to the sdk. `setTimeout` works here because it puts that next call in the queue

Reply:
@jparklev i came here because i got an email from github about the original comment text which mentioned me, and i didn't notice you removed the mention until after i dug into the code, so here's my 2 cents :)

basically the root cause of the race condition here is a break in the await/promise chain, at the point of calling `maker.addAccount`. @sirromdev you should be doing `await maker.addAccount` and not calling `useAccount`, changing the UI, etc. until after that. the button click should do absolutely nothing but call the choose callback with the selected address.

Reply:
i'm not sure `useMakerState` is the right fit here? because you never use `read` or `value`. this would make more sense to me:

```js
const { maker } = useMaker();
useEffect(async () => {
  const account = await maker.addAccount({ ... });
  // `account` arg has name, type, etc.
  // at this point a choice has been made and it's safe to call useAccount, etc
});
```

btw @jparklev, i see that one of the things `useMakerState` does is check authentication. note that [since v0.12.0](https://github.com/makerdao/dai.js/commit/10486e32af9e0ec98cdd493f58b66ce045f626d5) we can assume authentication has succeeded by the time `await Maker.create()` is done, unless `autoAuthenticate: false` was set

Reply:
https://github.com/makerdao/mcd-cdp-portal/pull/22#discussion_r269915445 😃